interfaces/connection.go

```diff
@@ -0,0 +1,245 @@
+// Connection represents a connection between ConnectedPlug and ConnectedSlot
+type Connection struct {
+	plug *ConnectedPlug
+	slot *ConnectedSlot
+}
+
+// ConnectedPlug represents a plug that is connected to a slot.
+type ConnectedPlug struct {
+	plugInfo     *snap.PlugInfo
+	staticAttrs  map[string]interface{}
+	dynamicAttrs map[string]interface{}
+}
+
+// ConnectedSlot represents a slot that is connected to a plug.
+type ConnectedSlot struct {
+	slotInfo     *snap.SlotInfo
+	staticAttrs  map[string]interface{}
+	dynamicAttrs map[string]interface{}
+}
+
+// NewConnectedSlot creates an object representing a connected slot.
+func NewConnectedSlot(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) *ConnectedSlot {
+	return &ConnectedSlot{
+		slotInfo:     slot,
+		staticAttrs:  copyAttributes(slot.Attrs),
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+// NewConnectedPlug creates an object representing a connected plug.
+func NewConnectedPlug(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) *ConnectedPlug {
+	return &ConnectedPlug{
+		plugInfo:     plug,
+		staticAttrs:  copyAttributes(plug.Attrs),
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+// Interface returns the name of the interface for this plug.
+func (plug *ConnectedPlug) Interface() string {
+	return plug.plugInfo.Interface
+}
+
+// Name returns the name of this plug.
+func (plug *ConnectedPlug) Name() string {
+	return plug.plugInfo.Name
+}
+
+// Snap returns the snap Info of this plug.
+func (plug *ConnectedPlug) Snap() *snap.Info {
+	return plug.plugInfo.Snap
+}
+
+// Apps returns all the apps associated with this plug.
+func (plug *ConnectedPlug) Apps() map[string]*snap.AppInfo {
+	return plug.plugInfo.Apps
+}
+
+// Hooks returns all the hooks associated with this plug.
+func (plug *ConnectedPlug) Hooks() map[string]*snap.HookInfo {
+	return plug.plugInfo.Hooks
+}
+
+// SecurityTags returns the security tags for this plug.
+func (plug *ConnectedPlug) SecurityTags() []string {
+	return plug.plugInfo.SecurityTags()
+}
+
+// StaticAttr returns a static attribute with the given key, or error if attribute doesn't exist.
+func (plug *ConnectedPlug) StaticAttr(key string) (interface{}, error) {
+	if val, ok := plug.staticAttrs[key]; ok {
+		return val, nil
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+// StaticAttrs returns all static attributes.
+func (plug *ConnectedPlug) StaticAttrs() map[string]interface{} {
+	return plug.staticAttrs
+}
+
+// Attr returns a dynamic attribute with the given name. It falls back to returning static
+// attribute if dynamic one doesn't exist. Error is returned if neither dynamic nor static
+// attribute exist.
+func (plug *ConnectedPlug) Attr(key string) (interface{}, error) {
+	if plug.dynamicAttrs != nil {
+		if val, ok := plug.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return plug.StaticAttr(key)
+}
+
+// Attrs returns all dynamic attributes of this plug.
+func (plug *ConnectedPlug) Attrs() map[string]interface{} {
```

Comment:
should this be called DynamicAttrs given that it returns only the dynamic attrs?

Reply:
Yes, that sounds sensible to me.. @niemeyer ?

Reply:
I agree with @pedronis here.

Reply:
I can't see a good use case for a dynamic map attribute, and I agree it's confusing that this is called Attrs but returning partial data and unlike Attr itself. When do we ever care about looking just at dynamic attributes?  It feels like if we have such a method, we'll use this and make an error since we'll assume the map is complete.

I suggest dropping the method for now and waiting for the use case.